File: chrome/android/javatests/src/org/chromium/chrome/browser/profiling_host/TestAndroidShim.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.profiling_host;

import org.chromium.base.annotations.MainDex;

/**
 * Provides direct access to test_android_shim, which in turn forwards to
 * ProfilingTestDriver. Only used for testing.
 */
@MainDex
public class TestAndroidShim {
    public TestAndroidShim() {
        mNativeTestAndroidShim = nativeInit();
    }

    /**
     *  When |dynamicallyStartProfiling| is true, the test harness is
     *  responsible for starting profiling for the relevant processes.
     *  When |pseudoStacks| is true, the stacks use trace-event based stacks
     *  rather than native stacks.
     */
    public boolean runTestForMode(String mode, boolean dynamicallyStartProfiling, String stackMode,
            boolean shouldSample, boolean sampleEverything) {
        return nativeRunTestForMode(mNativeTestAndroidShim, mode, dynamicallyStartProfiling,
                stackMode, shouldSample, sampleEverything);
    }

    /**
     * Clean up the C++ side of this class.
     * After the call, this class instance shouldn't be used.
     */
    public void destroy() {
        if (mNativeTestAndroidShim != 0) {
            nativeDestroy(mNativeTestAndroidShim);
            mNativeTestAndroidShim = 0;
        }
    }

    private long mNativeTestAndroidShim;
    private native long nativeInit();
    private native void nativeDestroy(long nativeTestAndroidShim);
    private native boolean nativeRunTestForMode(long nativeTestAndroidShim, String mode,
            boolean dynamicallyStartProfiling, String stackMode, boolean shouldSample,
            boolean sampleEverything);
}

File: chrome/browser/profiling_host/test_android_shim.cc
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/profiling_host/test_android_shim.h"

#include "base/android/jni_string.h"
#include "chrome/browser/profiling_host/profiling_process_host.h"
#include "chrome/browser/profiling_host/profiling_test_driver.h"
#include "jni/TestAndroidShim_jni.h"

using base::android::JavaParamRef;
using base::android::ScopedJavaLocalRef;

static jlong JNI_TestAndroidShim_Init(JNIEnv* env,
                                      const JavaParamRef<jobject>& obj) {
  TestAndroidShim* profiler = new TestAndroidShim(env, obj);
  return reinterpret_cast<intptr_t>(profiler);
}

TestAndroidShim::TestAndroidShim(JNIEnv* env, jobject obj) {}

TestAndroidShim::~TestAndroidShim() {}

void TestAndroidShim::Destroy(JNIEnv* env, const JavaParamRef<jobject>& obj) {
  delete this;
}

jboolean TestAndroidShim::RunTestForMode(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& obj,
    const base::android::JavaParamRef<jstring>& mode) {
  profiling::ProfilingTestDriver driver;
  profiling::ProfilingTestDriver::Options options;
  options.mode = profiling::ProfilingProcessHost::ConvertStringToMode(
      base::android::ConvertJavaStringToUTF8(mode));
  options.profiling_already_started = true;
  return driver.RunTest(options);
}
